### PR TITLE
BUG: Handle nonexistent end dates when resampling

### LINKED ISSUE #58380: BUG: NonExistentTimeError with resample("7D"), but only when DST day is not part of the date range
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

ts = pd.Series(1, pd.date_range("2024-04-19", "2024-04-20", tz="Africa/Cairo", freq="15min"))
ts.resample("7D").sum()
```

### Issue Description

Above fails with `pytz.exceptions.NonExistentTimeError: 2024-04-26 00:00:00` even though that date falls outside the date range.

Strangely, when the end date is *later*, and thus the problematic date *is* included in the result, it doesn't fail anymore:

```python
ts = pd.Series(1, pd.date_range("2024-04-19", "2024-04-27", tz="Africa/Cairo", freq="15min"))
ts.resample("7D").sum()
```
returns:
```
2024-04-19 00:00:00+02:00    672
2024-04-26 01:00:00+03:00     93
Freq: 7D, dtype: int64
```

### Expected Behavior

No failure and the correct result, namely:
```
2024-04-19 00:00:00+02:00    97
Freq: 7D, dtype: int64
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : d9cdd2ee5a58015ef6f4d15c7226110c9aab8140
python                : 3.10.4.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 10
Version               : 10.0.19045
machine               : AMD64
processor             : Intel64 Family 6 Model 140 Stepping 1, GenuineIntel
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : English_Belgium.1252
pandas                : 2.2.2
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 69.5.1
pip                   : 23.3.1
Cython                : None
pytest                : 7.4.2
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : 3.1.2
lxml.etree            : 5.2.1
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : 3.1.3
IPython               : 8.4.0
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : 1.3.8
dataframe-api-compat  : None
fastparquet           : 2023.2.0
fsspec                : 2024.3.1
gcsfs                 : None
matplotlib            : 3.7.5
numba                 : None
numexpr               : 2.10.0
odfpy                 : None
openpyxl              : 3.1.2
pandas_gbq            : None
pyarrow               : 15.0.2
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.13.0
sqlalchemy            : 2.0.28
tables                : None
tabulate              : 0.9.0
xarray                : 2024.3.0
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by rob-sil**
- [x] closes #58380
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v3.0.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/resample/test_datetime_index.py [hunk 1]
```diff
@@ -958,6 +958,20 @@ def _create_series(values, timestamps, freq="D"):
...
+    # GH 58380
+    ts = Series(
+        1,
+        date_range("2024-04-19", "2024-04-20", tz="Africa/Cairo", freq="15min"),
+    )
+
+    expected = Series([len(ts)], index=DatetimeIndex([ts.index[0]], freq="7D"))
+
+    result = ts.resample("7D").sum()
+    print(f"{result=}")
```
Comment (rhshadrach) [inline pandas/tests/resample/test_datetime_index.py:971]: Can you remove the print.

File: doc/source/whatsnew/v3.0.0.rst [hunk 2]
```diff
@@ -615,6 +615,7 @@ Groupby/resample/rolling
 - Bug in :meth:`DataFrameGroupBy.cumsum` where it did not return the correct dtype when the label contained ``None``. (:issue:`58811`)
 - Bug in :meth:`DataFrameGroupby.transform` and :meth:`SeriesGroupby.transform` with a reducer and ``observed=False`` that coerces dtype to float when there are unobserved categories. (:issue:`55326`)
 - Bug in :meth:`Rolling.apply` where the applied function could be called on fewer than ``min_period`` periods if ``method="table"``. (:issue:`58868`)
+- Bug in :meth:`Series.resample` could raise when the the date range ended shortly before DST. (:issue:`58380`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v3.0.0.rst:618]: Not just before DST, but more generally before a non-existent time, right?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: pandas/tests/resample/test_datetime_index.py (hunk 1 above)
Comment (rob-sil) [reply pandas/tests/resample/test_datetime_index.py]: Oops, sorry about that!

File: pandas/core/resample.py [hunk 3]
```diff
@@ -2466,7 +2466,7 @@ def _get_timestamp_range_edges(
         )
         if isinstance(freq, Day):
             first = first.tz_localize(index_tz)
-            last = last.tz_localize(index_tz)
+            last = last.tz_localize(index_tz, nonexistent="shift_forward")
```
[the remark this answers is not in the document]
Comment (rob-sil) [reply pandas/core/resample.py:2469]: I don't think `last` here is used as the last element in the series. Instead, `last` needs to be strictly after all the data points:
https://github.com/pandas-dev/pandas/blob/0fadaa9c9da39760e29e05012b14c8ce0782b5fe/pandas/_libs/lib.pyx#L893-L894

Shifting backwards should raise when running near a nonexistent hour, in the rare case where the nanosecond before is in the index:
```python
import pandas as pd

almost_a_day = pd.Timedelta(days=1) - pd.Timedelta(nanoseconds=1)
timestamp = pd.to_datetime("2024-04-25").tz_localize("Africa/Cairo")
ts = pd.Series(
    1,
    [timestamp, timestamp + almost_a_day],
)
ts.resample("1D", closed="right").sum()
```

However, shifting forward runs with the following output:
```
2024-04-24 00:00:00+02:00    1
2024-04-25 00:00:00+02:00    1
Freq: D, dtype: int64
```

(I'm not sure why it gives a bin on 2024-04-24.)

**Review by mroeschke — APPROVED** [review]
LGTM merge when ready @rhshadrach

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @rob-sil!

**Pull request merged**